### PR TITLE
Kw/support keyword args and preserve parameters

### LINKED ISSUE #6: Support for keyword args in Ruby 3
From Ruby 3 onwards, keyword args and hashes [can no longer be used interchangeably](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/). 
This leads to the issue, that this gem currently is not supporting memoizing methods with keyword arguments for Ruby 3.
In Ruby 2.7 you'll get the deprecation warning 
```
warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/path/to/file/with/memoized/method.rb:<line>: warning: The called method `_unmemoized_<method_name>' is defined here
``` 

In Ruby 3 this will produce an error.

Example method that will break in Ruby 3:

```
memoize def foo(keyword_arg:)
  p keyword_arg
end
```

### PR DESCRIPTION
**Pull request by mak-dunkelziffer**
Closes #6

### CODE REVIEW

File: lib/memoized/parameters.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        when :key
+          @key_params << [param_type, param_name]
+        when :keyrest
+          @keyrest_params << [param_type, param_name]
+        else raise "unknown parameter type"
+        end
+      end
+    end
+
+    def clear_params
```
Comment (mak-dunkelziffer) [inline lib/memoized/parameters.rb:28]: Wird noch inlined.

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -1,6 +1,9 @@
 require 'date'
 require 'timecop'
 require 'memoized'
+unless RUBY_VERSION == '2.5.3'
```
Comment (jakobscholz) [inline spec/spec_helper.rb]: Please use a real version comparison, like `if Gem::Version.new(RUBY_VERSION) > Gem::Version.new('2.5.3')`.

File: lib/memoized.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
+require 'memoized/parameters'
+
 module Memoized
+  UNIQUE = 42.freeze
```
Comment (jakobscholz) [inline lib/memoized.rb]: As this constant is only needed in the `Parameters`, how about moving it there?

File: lib/memoized/parameters.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        "5, 5, 5"
+      when :keyreq
+        "#{param_name}: 7"
+      when :key
+        "#{param_name}: 11"
+      when :keyrest
+        "**{first: 13, second: 13}"
+      else raise "unknown parameter type"
+      end
+    end
```
Comment (jakobscholz) [inline lib/memoized/parameters.rb]: I don't like having the test methods here in the production code. As the tests look to me, they could also be hard coded strings in the tests, which would maybe even make them better understandable. WDYT?

Comment (mak-dunkelziffer) [reply lib/memoized/parameters.rb]: Hard coding won't work. I think you missed the prop_check specs. (New kind of specs using a new Gem that I wanted to try out). It generates methods with many different parameter signatures dynamically.

I moved the methods that are only needed in tests out of `/lib` and into the `spec_helper.rb`. Not sure, if that's the right place, but at least they won't be loaded in the production code this way.

File: memoized.gemspec [hunk 5]
```diff
@@ -25,4 +25,5 @@ Gem::Specification.new do |s|
   s.add_development_dependency('rake', '~> 10.4.2')
   s.add_development_dependency('rspec', '~> 3.5.0')
   s.add_development_dependency('timecop', '~> 0.8.0')
+  s.add_development_dependency('prop_check', '~> 0.14.1')
```
Comment (dastra-mak) [inline memoized.gemspec:28]: In general you normally either define dev dependencies in the gemspec or in the Gemfile, not both. The Gemfile is missing the `gemspec` line for this to work. Could you clean this up please?

**Pull request merged**